### ISSUE TITLE
[Question] Is there a way to make a random agent but keeping everything else?

### ISSUE THREAD

**Issue by EloyAnguiano** [DLR-RM/stable-baselines3#1374]
### ❓ Question

Hi, I+m trying to make a random agent as a baseline of a model but I have lots of stuffs of evaluation on Callbacks and things like that. Is theere a way to put a Random agent as an Algorithm class that does not train at all and performs a random action but keeping every callback to evaluate as it is for a normal agent?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by qgallouedec** [comment]
Check this for a random agent https://github.com/DLR-RM/stable-baselines3/issues/1331#issuecomment-1431634372

The callbacks are only triggered at training time, so I don't really understand what kind of stuff you want to keep. Can you give a minimal example?

**Comment by EloyAnguiano** [comment]
My actual code looks like this:
```
    # AGENT
    agent = PPO("MultiInputPolicy",
                gym_env,
                **AGENT,
                policy_kwargs=POLICY,
                tensorboard_log=f"./sb3/results/{NAME}/",
                device=DEVICE,
                verbose=1)

    # CALLBACKS
    agent_path = f"sb3/agents/{NAME}"
    #rewards_names = ENV_CONFIG[DIFFICULTY]["no-opp"]["other_rewards"].keys()

    eval_callback = FullRewardsCallback(eval_env=deepcopy(gym_env),
                                        eval_freq=1e3, n_eval_episodes=5, 
                                        best_model_save_path=agent_path, 
                                        deterministic=args.eval_deterministic,
                                        log_path='./agents/logs/results')
    # TRAIN
    agent.learn(total_timesteps=STEPS, 
                callback=CallbackList([eval_callback]))
```

I want to change something to use it as a random agent (first I made a PPo with ent_coef = 10000.0, but it tries to train as well and uses a lot of computation)

**Comment by qgallouedec** [comment]
Why call `model.learn` if the agent has no weight to optimize?

**Comment by EloyAnguiano** [comment]
My environment is very different at each trajectory, so I would like to plot (With the FullRewardsCallback) each of the rewards into Tensorboard so I have to use learn, isn+t it?

**Comment by qgallouedec** [comment]
In tensorboard, we plot metrics against training steps. In your case, there is no training. So what would be the x-axis?

**Comment by EloyAnguiano** [comment]
The X axis should be the training steps as well, but I want to "mock" that training. That could let me see if some evaluation trajectories (I suspect that he environment changes Its difficulty along the trajectories) were easier to make with a Random agent and check the differences with the real one.

**Comment by qgallouedec** [comment]
> were easier to make with a Random agent and check the differences with the real one.

Sorry but I don't understand what you want to do.
It is indeed possible to take a random agent as a baseline. But it is only necessary to evaluate it once. In general, we represent this baseline with a horizontal line. Example:

![](https://assets-global.website-files.com/621e749a546b7592125f38ed/62273d864fecbb5cf935777d_Human%20Objectives%2009.svg)

A trajectory is the result of the interaction of the agent with the environment. If the agent changes, the trajectory changes as well. I don't understand what you mean by a "trajectory to make".

**Comment by araffin** [comment]
Side remark: if you want a fully random agent that behaves like PPO, set `n_epochs` and learning rate to zero.

**Comment by qgallouedec** [comment]
> Side remark: if you want a fully random agent that behaves like PPO, set `n_epochs` and learning rate to zero.

Note that it will not be a truly random agent, but rather a randomly initialized untrained agent.

**Comment by EloyAnguiano** [comment]
> > were easier to make with a Random agent and check the differences with the real one.
> 
> Sorry but I don't understand what you want to do. It is indeed possible to take a random agent as a baseline. But it is only necessary to evaluate it once. In general, we represent this baseline with a horizontal line. Example (here, x is the number of queries but it's the same thing with number training steps):
> 
> ![](https://camo.githubusercontent.com/5dd4b335a1f16ab76d710cce6144b28c06a6f7d55d46e72ce255827a68e2433a/68747470733a2f2f6173736574732d676c6f62616c2e776562736974652d66696c65732e636f6d2f3632316537343961353436623735393231323566333865642f3632323733643836346665636262356366393335373737645f48756d616e2532304f626a6563746976657325323030392e737667)
> 
> A trajectory is the result of the interaction of the agent with the environment. If the agent changes, the trajectory changes as well. I don't understand what you mean by a "trajectory to make".

This is indeed true, but I'm not using an standard enviornment. As I said, my suspect is that the difficulty changes along new trajectories, so a random agent could perform better in some steps

**Comment by qgallouedec** [comment]
So in your env the more episodes you generate, the more difficult it gets? So when you call `reset`, it's not a real reset, right?

**Comment by EloyAnguiano** [comment]
That is my hypothesis. The environment is not mine and uses time series stored data for developing each trajectory, and the beginning of each trajectory may determine the ease of it.

**Comment by qgallouedec** [comment]
I'm not sure what you mean by trajectory. In RL, a trajectory is the sequence of observations resulting from the interaction of an agent with an environment. Thus, two different agents will produce two different trajectories, even if the seed of the environment is the same.

> That is my hypothesis

it is not problematic that the difficulty is randomly defined at the beginning of the episode. **However**, it is important to check that it is a real reset, i.e. that the difficulty distribution does not change with the number of reset calls. Otherwise, it would be a really non-standard environment, and sb3 (nor any library to my knowledge) does not support it.
